packages/interledger/src/admin-api/schema.graphql

```diff
@@ -1,22 +1,437 @@
+
+  "Transfer between Interledger accounts"
+  transfer(
+    originAmount: Int!
+    originAccountId: ID!
+    destinationAccountId: ID!
+    destinationAmount: Int
+    autoCommit: Boolean
+    idempotencyKey: ID!
+  ): TransferMutationResponse
+
+  "Provision trustline"
+  provisionTrustline: ProvisionTrustlineMutationResponse
+
+  "Extend Trustline"
+  extendTrustline(
+    "The id of the trustline to extend."
+    trustlineId: ID!
+    "Amount of additional line of credit."
+    amount: Int!
+    "Automatically utilized and applied to the account balance."
+    autoApply: Boolean! = false
+  ): ExtendTrustlineMutationResponse
+
+  "Revoke trustline"
+  revokeTrustline(
+    "The id of the trustline to extend."
+    trustlineId: ID!
+    "Amount of additional line of credit."
+    amount: Int!
+  ): RevokeTrustlineMutationResponse
+
+  "Utilize trustline"
+  utilizeTrustline(
+    "The id of the trustline to extend."
+    trustlineId: ID!
+    "Amount of additional line of credit."
+    amount: Int!
+  ): UtilizeTrustlineMutationResponse
+
+  "Settle trustline"
+  settleTrustline(
+    "The id of the trustline to extend."
+    trustlineId: ID!
+    "Amount of additional line of credit."
+    amount: Int!
+    "Replenish the account's line of credit commensurate with the debt settled."
+    autoApply: Boolean! = true
+  ): SettleTrustlineMutationResponse
+
+  "Create webhook"
+  createWebhook(
+    "The id of the account to create the webhook for."
+    ilpAccountId: ID!
+  ): CreateWebhookMutationResponse
+
+  "Update webhook"
+  updateWebhook(
+    "The id of the webhook to update."
+    webhookId: ID!
+  ): UpdateWebhookMutationResponse
+
+  "Delete webhook"
+  deleteWebhook(
+    "The id of the webhook to delete."
+    webhookId: ID!
+  ): DeleteWebhookMutationResponse
+
+  "Create deposit"
+  createDeposit(
+    "The id of the account to create the deposit for."
+    ilpAccountId: ID!
+    "Amount of deposit."
+    amount: Int!
+  ): CreateDepositMutationResponse
+
+  "Create withdrawal"
+  createWithdrawal(
+    "The id of the account to create the withdrawal for."
+    ilpAccountId: ID!
+    "Amount of deposit."
+    amount: Int!
+  ): CreateWithdrawalMutationResponse
+
+  "Finalize pending withdrawal"
+  finalizePendingWithdrawal(
+    "The id of the pending withdrawal to finalize."
+    withdrawalId: ID!
+  ): FinalizePendingWithdrawalMutationResponse
+
+  "Rollback pending withdrawal"
+  rollbackPendingWithdrawal(
+    "The id of the pending withdrawal to rollback."
+    withdrawalId: ID!
+  ): RollbackPendingWithdrawalMutationResponse
+}
+
+type IlpAccountsConnection {
+  pageInfo: PageInfo!
+  edges: [IlpAccountEdge!]!
+}
+
+type IlpAccountEdge {
+  node: IlpAccount!
+  cursor: String!
+}
+
+type IlpAccount {
+  id: ID!
+  enabled: Boolean!
+  superAccount: IlpAccount
+  subAccounts(
+    "Paginating forwards: the cursor before the the requested page."
+    after: String
```

Comment:
can we create an input type for Pagination so that we don't need to keep re-defining this?

Reply:
We can, theoretically, but I don't think we should.

Firstly, the pagination algorithm we use ([GraphQL Cursor Connections Specification](https://relay.dev/graphql/connections.htm)) defines it this way. This is mostly to optimize for the clients side - making it easier to cache and fetch more data without necessarily having to program the logic each time. Of coarse this can be adapted to take in an input rather than individual arguments. [Relevant apollo docs](https://www.apollographql.com/docs/react/pagination/cursor-based/#relay-style-cursor-pagination). Example implementation: [Shopify pagination spec](https://shopify.dev/api/usage/pagination-graphql).

Secondly, it seems like standard practice to only use `input` objects for mutations, whereas queries usually have individual arguments. Example: [Github graphql](https://docs.github.com/en/graphql/reference/queries).

Lastly, a graphql input can't implement an interface. The fields on an input object type can themselves refer to input object types, but you can't mix input and output types in your schema. So you would have to redefine this anyway for each input:

```
type Query {
  ilpAccounts(input: IlpAccountsInput): IlpAccountsConnection!
}

input IlpAccountsInput {
  "Paginating forwards: the cursor before the the requested page."
  after: String
  "Paginating backwards: the cursor after the the requested page."
  before: String
  "Paginating forwards: The first **n** elements from the page."
  first: Int
  "Paginating backwards: The last **n** elements from the page."
  last: Int
}

```
Which just feels like unnecessary abstraction, and also more typing :keyboard: